### ISSUE TITLE
BITS error when built aside of mysql_async

### ISSUE THREAD

**Issue by nappa85** [CleverCloud/lapin#343]
Using both lapin and mysql_async in a project ends up in 43 build error like this:
```
error[E0034]: multiple applicable items in scope
   --> /home/marco/.cargo/registry/src/github.com-1ecc6299db9ec823/bitvec-0.19.4/src/field.rs:307:25
    |
307 |         check("load", len, M::BITS);
    |                               ^^^^ multiple `BITS` found
    |
note: candidate #1 is defined in the trait `BitMemory`
   --> /home/marco/.cargo/registry/src/github.com-1ecc6299db9ec823/bitvec-0.19.4/src/mem.rs:29:2
    |
29  |     const BITS: u8 = mem::size_of::<Self>() as u8 * 8;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait `IsNumber`
   --> /home/marco/.cargo/registry/src/github.com-1ecc6299db9ec823/funty-1.2.0/src/lib.rs:144:2
    |
144 |     const BITS: u32;
    |     ^^^^^^^^^^^^^^^^
help: disambiguate the associated constant for candidate #1
    |
307 |         check("load", len, BitMemory::BITS);
    |                            ^^^^^^^^^^^^^^^
help: disambiguate the associated constant for candidate #2
    |
307 |         check("load", len, IsNumber::BITS);
    |                            ^^^^^^^^^^^^^^
```

I cant' really understand what's happening here, lapin alone takes in bitvec v0.19.5, mysql_async alone takes in bitvec v0.22.3, but the two combined takes in bitvec v0.22.3 AND v0.19.4 that isn't patched against BITS, introduced with rust 1.53.0

I'm opening the issue here because the problem seems to origin from this crate dependencies

**Comment by biot023** [comment]
+1 from me -- I am falling *seriously* afoul of this issue with the latest `lapin` and `mysql` crates.

**Comment by biot023** [comment]
Incidentally, there are instructions on how to replicate my issue, here: https://github.com/blackbeam/rust-mysql-simple/issues/289

**Comment by nappa85** [comment]
I solved it, but not in a clean and sustainable way:
git clone bitvec,  funty,  memchr  and nom at the desider versions, change the package name of bitvec, funty and memchr to something like "{}-temp", then edit nom/Cargo.toml to include it from local path and add a package key pointing to the new name, this way the code will see it with the old name.
At the end, add this to your Cargo.toml:
```
[patch.crates-io]
nom = { path = "./nom" }
```

This way you forked the four crates, changing the name so the crate resolver doesn't try to merge them with other versions.

I repeat, it isn't a sustainable solution, only a temporary one.

**Comment by Keruspe** [comment]
@Geal regression in nom?

Worked fine a couple of weeks ago

**Comment by nappa85** [comment]
> @Geal regression in nom?
> 
> Worked fine a couple of weeks ago

It's not a nom regression, it' s a problem with funty, version 1.2 isn't compatible with previous, and nom is version-locked to funty > 1.0 <= 1.1, mysql_async (and I think many other crates) have already updated to funty 1.2, nom in the next developing version seems to have dropped the funty dependency

**Comment by Geal** [comment]
yes, that's why I'm trying to finish nom's 7.0 release fast

**Comment by Keruspe** [comment]
No way of fixing this for nom 6?

**Comment by Keruspe** [comment]
> > @Geal regression in nom?
> 
> > 
> 
> > Worked fine a couple of weeks ago
> 
> 
> 
> It's not a nom regression, it' s a problem with funty, version 1.2 isn't compatible with previous, and nom is version-locked to funty > 1.0 <= 1.1, mysql_async (and I think many other crates) have already updated to funty 1.2, nom in the next developing version seems to have dropped the funty dependency

Yes, so the dependency nom->funty is what broke. Not something lapin is doing directly

**Comment by nappa85** [comment]
> Yes, so the dependency nom->funty is what broke. Not something lapin is doing directly

Yes, sorry, for having opened the issue here, I've deeply analyzed the problem in the mean time, I've opened this issue 8 days ago...

**Comment by Keruspe** [comment]
> > Yes, so the dependency nom->funty is what broke. Not something lapin is doing directly
> 
> 
> 
> Yes, sorry for having opened the issue here, I've deeply analyzed the problem in the mean time, I've opened this issue 8 days ago...

I think it's a right place for this issue, given the "fix" might be a switch to nom 7

**Comment by nappa85** [comment]
Maybe nom6 could hard-fork funty 1.1 under a different name, as a temporary solution.
It could be a solution for other packages too.
On funty crates many people are asking to trash the 1.2 version and republish it as 2.0.
I haven't understood the rules that the compiler uses to build two different versions of the same crate

**Comment by Keruspe** [comment]
Fixed in 1.8.0
